Update assembly version from hardcoded to MajorVersion and update NetCoreAppCurrent to net8.0

carlossanlop opened the pull request:
Contributes to https://github.com/dotnet/runtime/issues/72112
Fixes https://github.com/dotnet/runtime/issues/58109

Co-authored with @radical and @ViktorHofer 

Original PR was https://github.com/dotnet/runtime/pull/74157 but we had to revert it https://github.com/dotnet/runtime/pull/77899

Reason why it had to be reverted:

> _Changing assembly versions without changing the TargetFramework represented by those assemblies is causing any consuming assembly in packages that target net7.0 to have references to 8.0 versioned assemblies. This will break the consumption of those packages._

> _For this change to work, it cannot change referenced versions that are exposed as net7.0. The TFM targeted by packages can likely change before AssemblyVersion, but not vice-versa._

carlossanlop committed:
Update assembly version from hardcoded to MajorVersion

carlossanlop committed:
Adjust Microsoft.VisualBasic.Core

carlossanlop committed:
Adjust System.Diagnostics.EventLog

carlossanlop committed:
Adjust System.Resources.Extensions

carlossanlop committed:
Adjust System.Runtime.Serialization.Xml

carlossanlop committed:
Extract the LocalEchoServer bits from the projects so that the helix bits are in a LocalEchoServer.helix.targets, instead of being special-cased in sendtohelix-wasm.targets. And this is setup and used by the test projects by importing one file.

carlossanlop committed:
Build the LocalEchoServers specifically without using artifacts

- The RemoteLoopServer, and NetCoreServer are projects used as aspnetcore middleware, and loaded by xharness.
- These are built against the live artifacts, same as other projects.
    - But this can be a problem when the libraries in `runtime` are on a newer assembly version (say `8.0.0`), but xharness is still built with `7.0.0` libraries.
    - In that case, xharness fails to load the middleware:

```
Application startup exception: System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types.

...

Could not load file or assembly 'System.Runtime, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.

   at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
   at System.Reflection.Assembly.GetTypes()
   at Microsoft.DotNet.XHarness.CLI.CommandArguments.TypeFromAssemblyArgument`1.GetLoadedTypes()+MoveNext() in /_/src/Microsoft.DotNet.XHarness.CLI/CommandArguments/Arguments/TypeFromAssemblyArgument.cs:line 29
   at Microsoft.DotNet.XHarness.CLI.Commands.WebServer.<>c__DisplayClass0_0.<Start>b__9(TestWebServerOptions options) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/WebServer.cs:line 60
   at Microsoft.Extensions.Options.OptionsFactory`1.Create(String name)
   at Microsoft.Extensions.Options.OptionsCache`1.<>c__3`1.<GetOrAdd>b__3_0(String name, ValueTuple`2 arg)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd[TArg](TKey key, Func`3 valueFactory, TArg factoryArgument)
   at Microsoft.Extensions.Options.OptionsCache`1.GetOrAdd[TArg](String name, Func`3 createOptions, TArg factoryArgument)
   at Microsoft.DotNet.XHarness.CLI.Commands.WebServer.TestWebServerStartup.Configure(IApplicationBuilder app, IOptionsMonitor`1 optionsAccessor) in /_/src/Microsoft.DotNet.XHarness.CLI/Commands/WebServer.cs:line 126
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.AspNetCore.Hosting.ConfigureBuilder.Invoke(Object instance, IApplicationBuilder builder)
   at Microsoft.AspNetCore.Hosting.ConventionBasedStartup.Configure(IApplicationBuilder app)
   at Microsoft.AspNetCore.Hosting.WebHost.BuildApplication()
```

- Build the project in isolation from rest of the repo, so that it is built with references only from the SDK. The built assembly can then be deployed for use with xharness, just like before.

carlossanlop committed:
Add missing files

carlossanlop committed:
Use eng/targetingpacks.targets instead of wasm specific targets

carlossanlop committed:
Avoid duplicate imports, and don't use the targetingpacks.targets for runtime tests since the paths are set explicitly in the project

ViktorHofer committed:
Update NetCoreAppCurrent to net8.0 and NetCoreAppPrevious

ViktorHofer committed:
Add NetCoreAppPrevious tfm to out-of-band projects

carlossanlop commented on Directory.Build.props:
```diff
@@ -46,23 +46,27 @@
   <!-- The TFMs to build and test against. -->
   <PropertyGroup>
-    <NetCoreAppCurrentVersion>7.0</NetCoreAppCurrentVersion>
+    <NetCoreAppCurrentVersion>8.0</NetCoreAppCurrentVersion>
```
Shouldn't we be using the values we have in eng/Versions.props that we added in this PR? https://github.com/dotnet/runtime/pull/78354/commits/17764b999e7dfece2e72ca1bf7c68aa73465c8cd

Here, and also elsewhere where we hardcode the minor or major numbers.

ViktorHofer replied:
Directory.Build.props has access to Versions.props but we don't want to use the version properties from there as that would complicate branding to .NET 9, as it would mean that the package version branding change would also rev the NetCoreAppCurrent version, and we don't want that.

carlossanlop commented on src/libraries/System.Data.Odbc/src/System.Data.Odbc.csproj:
```diff
@@ -1,6 +1,6 @@
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-FreeBSD;$(NetCoreAppCurrent)-illumos;$(NetCoreAppCurrent)-Solaris;$(NetCoreAppCurrent)-Linux;$(NetCoreAppCurrent)-OSX;$(NetCoreAppCurrent)-iOS;$(NetCoreAppCurrent)-tvOS;$(NetCoreAppCurrent);$(NetCoreAppMinimum)-windows;$(NetCoreAppMinimum)-FreeBSD;$(NetCoreAppMinimum)-Linux;$(NetCoreAppMinimum)-OSX;$(NetCoreAppMinimum);netsta ...
+    <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-FreeBSD;$(NetCoreAppCurrent)-illumos;$(NetCoreAppCurrent)-Solaris;$(NetCoreAppCurrent)-Linux;$(NetCoreAppCurrent)-OSX;$(NetCoreAppCurrent)-iOS;$(NetCoreAppCurrent)-tvOS;$(NetCoreAppCurrent);$(NetCoreAppPrevious)-windows;$(NetCoreAppPrevious)-FreeBSD;$(NetCoreAppPrevious)-illumos;$(NetCoreAppPrevious)-Solaris;$(NetCoreAppPrevi ...
```
Fun! 😄

carlossanlop commented on Directory.Build.props:
```diff
@@ -46,23 +46,27 @@
+    <NetCoreAppPreviousVersion>7.0</NetCoreAppPreviousVersion>
+    <NetCoreAppPrevious>net$(NetCoreAppPreviousVersion)</NetCoreAppPrevious>
+    <NetCoreAppPrevious Condition="'$(DotNetBuildFromSource)' == 'true'">$(NetCoreAppCurrent)</NetCoreAppPrevious>
+    <NetCoreAppPreviousPackageBaselineVersion>$(NetCoreAppPreviousVersion).0</NetCoreAppPreviousPackageBaselineVersion>
```
Is this correct? This would become a `7.0.0`

ViktorHofer replied:
Yes, ApiCompat restores the previous stable package from nuget, which now is 7.0.0.

carlossanlop commented:
@ViktorHofer do we need any other changes to prevent breaking the other repos, or are we good now?

I can't approve this PR because I submitted it, and I assume @ViktorHofer can't either since he contributed to it too.

@agocke / @ericstj / @smasher164 / @radical can we get a sign-off from at least one of you?

ViktorHofer commented:
@carlossanlop we just started working on this yesterday and we have another meeting tomorrow in which we will continue to make progress. This PR requires many more changes (I expect another 50-100 file changes). Also, the build is failing because we haven't yet updated all the places that require updates.

ViktorHofer committed:
Update reference source projects with new tfm

ViktorHofer committed:
Set maximum version early enough for SDK features

ViktorHofer committed:
Update suppression files for tfm change

ViktorHofer committed:
Merge remote-tracking branch 'origin/main' into Net8ChangesSecondAttempt

ViktorHofer committed:
Fix libraries build because of ILLink warning

ViktorHofer committed:
Update 'net7.0' usages in the repo

ViktorHofer committed:
Update ymls

ViktorHofer committed:
Make trimming tests use the live built apphost

Fixes https://github.com/dotnet/runtime/issues/58109

ViktorHofer committed:
Add trimmingTests.props

ViktorHofer committed:
UseLiveAppHost

ViktorHofer committed:
Merge remote-tracking branch 'upstream/main' into Net8ChangesSecondAttempt

ViktorHofer committed:
Clean-up shared CoreLib suppression file

ViktorHofer committed:
Live apphost issues with NativeExports

ViktorHofer committed:
Merge branch 'main' into Net8ChangesSecondAttempt

ViktorHofer committed:
Merge branch 'main' into Net8ChangesSecondAttempt

ViktorHofer commented:
> /__w/1/s/.dotnet/sdk/7.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(267,5): error NETSDK1005: Assets file '/__w/1/s/artifacts/tests/coreclr/packages/Common/test_dependencies/test_dependencies/project.assets.json' doesn't have a target for 'net8.0'. Ensure that restore has run and that you have included 'net8.0' in the TargetFrameworks for your project. [/__w/1/s/src/tests/JIT/Regression/JitBlue/Runtime_60035/Runtime_60035.csproj] [/__w/1/s/src/tests/build.proj]
##[error].dotnet/sdk/7.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(267,5): error NETSDK1005: Assets file '/__w/1/s/artifacts/tests/coreclr/packages/Common/test_dependencies/test_dependencies/project.assets.json' doesn't have a target for 'net8.0'. Ensure that restore has run and that you have included 'net8.0' in the TargetFrameworks for your project. [/__w/1/s/src/tests/JIT/Regression/JitBlue/Runtime_60035/Runtime_60035.csproj]

@jkoritzinsky @hoyosjs @agocke can you please look into the failure here? Should test_dependencies.csproj/fsproj target `$(NetCoreAppToolCurrent)` or `$(NetCoreAppCurrent)`? This PR updates `$(NetCoreAppCurrent)` to `net8.0`.

ViktorHofer committed:
Update hardcoded tfm values in wasm build files

ViktorHofer commented:
> C:\Program Files\dotnet\sdk\7.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(90,5): error NETSDK1183: Unable to optimize assemblies for Ahead of time compilation: a valid runtime package was not found. Either set the PublishAot property to false, or use a supported runtime identifier when publishing. When targeting .NET 7 or higher, make sure to restore packages with the PublishAot property set to true. [D:\a\_work\1\s\artifacts\bin\trimmingTests\projects\System.Diagnostics.DiagnosticSource.NativeAotTests\DiagnosticSourceEventSourceTests\win-x64\project.csproj]

@MichalStrehovsky do you know who can help with that error? This PR upgrades the tfm from net7.0 to net8.0 and I'm seeing the linker tests now failing. That's from the `windows x64 release NativeAOT` leg.

agocke commented:
I’m looking at the test dependencies. @sbomer  can you look at the Native aot error? Might be the same warning problem

radical committed:
[wasm] Fix debugger tests to run with net8.0

agocke committed:
Remove explicit TF specification from JIT test

radical committed:
Merge branch 'Net8ChangesSecondAttempt' of https://github.com/carlossanlop/runtime into Net8ChangesSecondAttempt

radical committed:
[wasi] Sample: change net7.0 -> net8.0

This is temporary, and the whole sample is changing very soon with the
on-going wasi work.

radical committed:
add comment

radical commented:
- wasm/debugger tests should be fixed now.
- And the failing wasi build on `LibraryTests` job should work too

sbomer committed:
Don't try to restore 8.0 ILC with 7.0 SDK

radical commented:
Wasm/debugger-tests/windows will be fixed by https://github.com/dotnet/runtime/pull/79124 .

Wasm threading, and perforating builds are [failing with](https://dev.azure.com/dnceng-public/public/_build/results?buildId=99596&view=logs&jobId=1567862a-6a9e-5016-9353-cc84f1411dc4&j=8d0bfc3e-d1bc-5879-4cc4-18653f9adb55&t=d60309cd-b93c-533c-9d7c-31b2b5b95034):

```
  API compatibility errors between 'ref/net8.0/System.Diagnostics.Tracing.dll' (left) and 'lib/net8.0/System.Diagnostics.Tracing.dll' (right):
error CP0014: Cannot remove attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")' from 'System.Diagnostics.Tracing.DiagnosticCounter'. [/__w/1/s/src/libraries/System.Diagnostics.Tracing/src/System.Diagnostics.Tracing.csproj]
error CP0014: Cannot remove attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")' from 'System.Diagnostics.Tracing.EventCounter'. [/__w/1/s/src/libraries/System.Diagnostics.Tracing/src/System.Diagnostics.Tracing.csproj]
error CP0014: Cannot remove attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")' from 'System.Diagnostics.Tracing.IncrementingEventCounter'. [/__w/1/s/src/libraries/System.Diagnostics.Tracing/src/System.Diagnostics.Tracing.csproj]
error CP0014: Cannot remove attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")' from 'System.Diagnostics.Tracing.IncrementingPollingCounter'. [/__w/1/s/src/libraries/System.Diagnostics.Tracing/src/System.Diagnostics.Tracing.csproj]
error CP0014: Cannot remove attribute 'System.Runtime.Versioning.UnsupportedOSPlatformAttribute("browser")' from 'System.Diagnostics.Tracing.PollingCounter'. [/__w/1/s/src/libraries/System.Diagnostics.Tracing/src/System.Diagnostics.Tracing.csproj]
```

cc @pavelsavara @lambdageek

radical committed:
Add missing file

radical commented:
And I pushed a change to fix the wasm library tests build.

ViktorHofer committed:
Add KnownILCompilerPack entry to targetingpacks.targets

ViktorHofer commented:
> A note about the workaround: I tried to work around the missing KnownILCompilerPack by instead adding it in targetingpacks.targets, and then removing items from PackageDownload. This wasn't working because the nativeaot tests aren't restored in a standard way:

Thanks for providing the workaround @sbomer. I just tried your original approach out and it worked for me. I removed your workaround and added a `KnownILCompilerPack` entry instead. One general note, now that the libraries linker and aot tests use a life apphost, I had to add the `clr.runtime+host.native` subset to the nativeaot build lanes. I just wanted to build the 
apphost via the `host.native` subset but that didn't work without building the singlefilehost first. I opened https://github.com/dotnet/runtime/issues/79144 for that.

ViktorHofer committed:
Merge remote-tracking branch 'origin/main' into Net8ChangesSecondAttempt

ViktorHofer committed:
Fix Hashing API only available in ref/Hashing net8.0.

ViktorHofer committed:
Fix APICompat issue

ViktorHofer committed:
Fix more apicompat issues

ViktorHofer committed:
Specify host configuration when building the host in yml

ViktorHofer committed:
Update AppleApp local build NetCoreAppCurrent

ViktorHofer commented:
Who could help with the DNNE error? I initially switched it to use the live app host but that resulted in a dependency on the host subset when building the libraries tests which we want to avoid (with our current loosely coupled subset system). Now it again uses the SDK host package that should be available, but I have no idea why it's failing to compile.

```
  Building native export: "/usr/bin/clang-9" -O2 -shared -fpic -D DNNE_ASSEMBLY_NAME=Microsoft.Interop.Tests.NativeExports -D DNNE_COMPILE_AS_SOURCE -I "/__w/1/s/.packages/dnne/1.0.27/tools/platform" -I "/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native" -o "/__w/1/s/artifacts/obj/NativeExports/Release/net8.0/linux-musl-arm/dnne/bin/Microsoft.Interop.Tests.NativeExportsNE.so" --target=armv7-alpine-linux-musleabihf --gcc-toolchain=/crossrootfs/arm/usr --sysroot=/crossrootfs/arm -mthumb -mfpu=vfpv3 -DCLR_ARM_FPU_CAPABILITY=0x7 "/__w/1/s/artifacts/obj/NativeExports/Release/net8.0/linux-musl-arm/dnne/Microsoft.Interop.Tests.NativeExports.g.c" "/__w/1/s/.packages/dnne/1.0.27/tools/platform/platform.c" -lstdc++ "/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native/libnethost.a" --target=armv7-alpine-linux-musleabihf --gcc-toolchain=/crossrootfs/arm/usr --sysroot=/crossrootfs/arm -fuse-ld=lld  -Wl,--rpath-link=/crossrootfs/arm/lib/armv7-alpine-linux-musleabihf -Wl,--rpath-link=/crossrootfs/arm/usr/lib/armv7-alpine-linux-musleabihf 
clang : error : no such file or directory: '/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native/libnethost.a' [/__w/1/s/src/libraries/System.Runtime.InteropServices/tests/TestAssets/NativeExports/NativeExports.csproj]
##[error]clang(0,0): error : (NETCORE_ENGINEERING_TELEMETRY=Build) no such file or directory: '/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native/libnethost.a'
/__w/1/s/.packages/dnne/1.0.27/build/DNNE.targets(145,5): error MSB3073: The command ""/usr/bin/clang-9" -O2 -shared -fpic -D DNNE_ASSEMBLY_NAME=Microsoft.Interop.Tests.NativeExports -D DNNE_COMPILE_AS_SOURCE -I "/__w/1/s/.packages/dnne/1.0.27/tools/platform" -I "/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native" -o "/__w/1/s/artifacts/obj/NativeExports/Release/net8.0/linux-musl-arm/dnne/bin/Microsoft.Interop.Tests.NativeExportsNE.so" --target=armv7-alpine-linux-musleabihf --gcc-toolchain=/crossrootfs/arm/usr --sysroot=/crossrootfs/arm -mthumb -mfpu=vfpv3 -DCLR_ARM_FPU_CAPABILITY=0x7 "/__w/1/s/artifacts/obj/NativeExports/Release/net8.0/linux-musl-arm/dnne/Microsoft.Interop.Tests.NativeExports.g.c" "/__w/1/s/.packages/dnne/1.0.27/tools/platform/platform.c" -lstdc++ "/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native/libnethost.a" --target=armv7-alpine-linux-musleabihf --gcc-toolchain=/crossrootfs/arm/usr --sysroot=/crossrootfs/arm -fuse-ld=lld  -Wl,--rpath-link=/crossrootfs/arm/lib/armv7-alpine-linux-musleabihf -Wl,--rpath-link=/crossrootfs/arm/usr/lib/armv7-alpine-linux-musleabihf " exited with code 1. [/__w/1/s/src/libraries/System.Runtime.InteropServices/tests/TestAssets/NativeExports/NativeExports.csproj]
##[error].packages/dnne/1.0.27/build/DNNE.targets(145,5): error MSB3073: (NETCORE_ENGINEERING_TELEMETRY=Build) The command ""/usr/bin/clang-9" -O2 -shared -fpic -D DNNE_ASSEMBLY_NAME=Microsoft.Interop.Tests.NativeExports -D DNNE_COMPILE_AS_SOURCE -I "/__w/1/s/.packages/dnne/1.0.27/tools/platform" -I "/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native" -o "/__w/1/s/artifacts/obj/NativeExports/Release/net8.0/linux-musl-arm/dnne/bin/Microsoft.Interop.Tests.NativeExportsNE.so" --target=armv7-alpine-linux-musleabihf --gcc-toolchain=/crossrootfs/arm/usr --sysroot=/crossrootfs/arm -mthumb -mfpu=vfpv3 -DCLR_ARM_FPU_CAPABILITY=0x7 "/__w/1/s/artifacts/obj/NativeExports/Release/net8.0/linux-musl-arm/dnne/Microsoft.Interop.Tests.NativeExports.g.c" "/__w/1/s/.packages/dnne/1.0.27/tools/platform/platform.c" -lstdc++ "/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native/libnethost.a" --target=armv7-alpine-linux-musleabihf --gcc-toolchain=/crossrootfs/arm/usr --sysroot=/crossrootfs/arm -fuse-ld=lld  -Wl,--rpath-link=/crossrootfs/arm/lib/armv7-alpine-linux-musleabihf -Wl,--rpath-link=/crossrootfs/arm/usr/lib/armv7-alpine-linux-musleabihf " exited with code 1.
```

cc @AaronRobinsonMSFT

ViktorHofer committed:
Fix S.S.Permissions apicompat suppression file

ViktorHofer commented:
The installer tests are failing because of https://github.com/dotnet/runtime/issues/45972.

ViktorHofer committed:
Hack support for net8.0 tfm in installer tests

AaronRobinsonMSFT commented:
> no such file or directory: '/__w/1/s/.dotnet/packs/Microsoft.NETCore.App.Host.linux-musl-arm/7.0.0/runtimes/linux-musl-arm/native/libnethost.a

This seems like the root issue. Is this file there? Is the path to the musl lib different in this case? I've no idea why this would be failing, but do have an arm64 macOS machine so will try to repro locally.

AaronRobinsonMSFT commented:
Starting from a clean repo (i.e., `git clean -xdf`) on an arm64-macOS I ran the following command:

`./build.sh -subset libs+libs.tests   -configuration Debug -arch arm64 -framework net8.0 -testscope innerloop`

This is the same command in the above failing macOS leg. I am thinking this is a race condition with the file system and restore because this works on other legs without issue and macOS nor linux is special cased in any way with respect to command execution. I am going to re-run the failing legs again to test this hypothesis.

The MSBuild target in DNNE for this area can be found at https://github.com/AaronRobinsonMSFT/DNNE/blob/236974815bbefd80d844d9c38f8fb154df33684b/src/msbuild/DNNE.targets#L106-L157. It is very simple and I've no idea how anything in there could be causing the failure only on musl and arm64-macos.

ViktorHofer commented:
> This is the same command in the above failing macOS leg. I am thinking this is a race condition with the file system and restore because this works on other legs without issue and macOS nor linux is special cased in any way with respect to command execution. I am going to re-run the failing legs again to test this hypothesis.

I think this is more about legs that cross-build, as only such fail on CI. Except for FreeBSD but apparently NativeExports isn't compiled on that leg at all.

ViktorHofer committed:
Allow NativeExports to use a prebuilt apphost

ViktorHofer reviewed (approved):
Approving others and my own changes.

AaronRobinsonMSFT commented:
@ViktorHofer Can I do anything in DNNE to make this better?

ViktorHofer committed:
UseLocalAppHostPack property

agocke commented:
> > A note about the workaround: I tried to work around the missing KnownILCompilerPack by instead adding it in targetingpacks.targets, and then removing items from PackageDownload. This wasn't working because the nativeaot tests aren't restored in a standard way:
> 
> Thanks for providing the workaround @sbomer. I just tried your original approach out and it worked for me. I removed your workaround and added a `KnownILCompilerPack` entry instead. Related, now that the libraries linker and aot tests use the live apphost, I had to add the `clr.runtime+host.native` subset to the nativeaot build lanes. I just wanted to build the apphost via the `host.native` subset but that didn't work without building the singlefilehost first. I opened #79144 for that.

I don't want to block this PR, but consider just redefining `clr.aot` to be whatever you had to write to get this to work. The `clr.aot` partition is meant to be "whatever's needed for AOT in common cases", not a specific set of stuff.

ViktorHofer committed:
Fix apphostsourcepath in DNNE project

ViktorHofer commented:
> I don't want to block this PR, but consider just redefining clr.aot to be whatever you had to write to get this to work. The clr.aot partition is meant to be "whatever's needed for AOT in common cases", not a specific set of stuff.

That sounds like a good strategy until we have [strongly defined dependencies in the repo](https://github.com/dotnet/runtime/issues/43110). But before adding the `clr.runtime` subset to `clr.aot` we should first resolve the discussion in https://github.com/dotnet/runtime/issues/79144 as I don't think that the `clr.runtime` subset is necessary at all but makes the AOT build significantly slower.

ViktorHofer commented:
> @ViktorHofer Can I do anything in DNNE to make this better?

DNNE's msbuild integration is well written and allows to customize any of the used properties. I doubt that we would need improvements there to simplify our build. The problem is dotnet/runtime itself which makes it hard to depend on live built artifacts (like the host) because of the way it currently builds in CI (runtime and libraries in parallel on separate machines).

MichalStrehovsky commented:
> That sounds like a good strategy until we have https://github.com/dotnet/runtime/issues/43110. But before adding the clr.runtime subset to clr.aot we should first resolve the discussion in https://github.com/dotnet/runtime/issues/79144 as I don't think that the clr.runtime subset is necessary at all and makes the AOT build significantly slower.

Agreed, especially if the host is only needed to run the `RunNativeAotTestApps` mode introduced in #76109. This is not a set of tests a runtime developer would run - they guard libraries invariants and are not interesting for runtime devs. I really wish we did something so that the host is not looked for in that mode (not in this PR I guess). I don't know why it's looked for.

MichalStrehovsky commented on docs/workflow/building/coreclr/nativeaot.md:
```diff
@@ -101,7 +101,7 @@ For more advanced scenarios, look for at [Building the Tests](/docs/workflow/tes
 ### Running library tests
 
-Build library tests by passing the `libs.tests` subset together with the `/p:TestNativeAot=true` to build the libraries, i.e. `clr.aot+libs+libs.tests /p:TestNativeAot=true` together with the full arguments as specified [above](#building). Then, to run a specific library, go to the tests directory of the library and run the usual command to run tests for the library (see [Running tests for a sing ...
+Build library tests by passing the `libs.tests` subset together with the `/p:TestNativeAot=true` to build the libraries, i.e. `clr.runtime+clr.aot+host.native+libs+libs.tests /p:TestNativeAot=true` together with the full arguments as specified [above](#building). Then, to run a specific library, go to the tests directory of the library and run the usual command to run tests for the library (see [ ...
```
Is this really necessary?

I thought the failures we saw were related to the `RunNativeAotTestApps` step that was added in #76109. We don't run that in this doc.

ViktorHofer replied:
Thanks, you are right. I mixed up `RunNativeAotTestApps` and `TestNativeAot`. Will send a follow-up to revert this change to not block this PR.
